text/0374-jsii-ts-version.md

```diff
@@ -0,0 +1,324 @@
+
+In order to allow each package author to decide for themselves without hinging
+on their dependencie's choices, or influencing their dependents, we are making
+it possible for every developer to decide when they want to migrate to a new
+version of the [TypeScript] language, by following [TypeScript]'s versioning for
+the compiler, and continuing to perform coordinated releases of every other
+package in the toolchain.
+
+### Why should we _not_ do this?
+
+[Semantic versioning][semver] is the dominant versioning scheme in the
+JavaScript world, and diverging from it might break assumptions customers make
+when consuming packages from [npm](npmjs.com).
+
+Following [TypeScript] versions means releasing new `major.minor` releases (that
+may include breaking changes) more often than is currently done (the
+[TypeScript] maintainers typically declare a new `major.minor` line four
+times per year: in February, May, August and November respectively). This is a
+lot more release lines than other AWS products offer, and it would not be
+reasonable to uphold the [AWS SDKs and Tools maintenance policy][aws-policy],
+as it requires offering full support (features and bug fixes) for "Generally
+Available" releases for a minimum of 24 months. Diverging from this
+[policy][aws-policy] is a significant change that customers may not expect.
+
+### What is the technical solution (design) of this feature?
+
+The proposed delivery plan for this feature is as follows:
+
+* Communicate about the upcoming change in new releases on the `v1` release
+  line, including a planned timeline for the initial release of the `v4.7` line,
+  and language about the departure from [SemVer].
+
+* **Optional:** Separate the `jsii` compiler from the rest of packages in the
+  jsii toolchain into separate repositories. This would make it easier to
+  release the `jsii` compiler separately from other parts of the toolchain that
+  follow a different versioning scheme.
+
+  - Move other packages to a new mono-repository (or to several new individual
+    package repositories), e.g: `github.com/aws/jsii-toolchain`.
+
+  - Make the `github.com/aws/jsii` repository by a single-package repository.
+
+* Update the *release* automation to use the [TypeScript] compiler version's
+  `major.minor` level, and only update the `patch` level on new releases.
+
+* Upgrade the [TypeScript] compiler internally used by `jsii` to the current
+  `latest` release of the `typescript` package.
+
+* Make the `jsii` compiler transparently prepare down-leveled declarations files
+  to ensure backwards-compatibility of compiler outputs with previous releases
+  of `jsii`. This can be done using the [`downlevel-dts`][downlevel-dts] package
+  to produce declarations files compatible with [TypeScript] `3.9`, and adding
+  a `typesVersions` key to the `package.json` file of packages.
+
+  - See proof-of-concept: [`aws/jsii#3501`][aws/jsii-3501]
+
+* Release the initial `jsii` release on the `4.7` line.
```

Comment:
So we are choosing to disregard all versions between 3.9 and 4.7, yeah?

Reply:
Yes. I don't think there is much of a point in inflicting that additional work on us. That is, unless it turns out your "adapter" idea is feasible (then I guess why not).